File: 0803-cheapest-flights-within-k-stops/solution.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        edges = collections.defaultdict(list)
        for x, v, w in flights:
            edges[x].append((v, w))
        heap = [(0, src, 0)]
        visited = [float('inf')] * n
        while heap:
            weight, val, count = heapq.heappop(heap)
            if count > k + 1:
                continue
            if count >= visited[val]:
                continue
            if val == dst:
                return weight
            visited[val] = count
            for val1, weight1 in edges[val]:
                heapq.heappush(heap, (weight1 + weight, val1, count + 1))
        return -1
                                

File: 0803-cheapest-flights-within-k-stops/solution.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        vector<int> prices(n + 1, INT_MAX);
        prices[src] = 0;
        for(int i = 0; i < k + 1; ++i){
            vector<int> tempPrices = prices;
            for(const auto& cur : flights){
                int from = cur[0];
                int to = cur[1];
                int price = cur[2];
                if(prices[from] == INT_MAX){
                    continue;
                }
                if(prices[from] + price < tempPrices[to]){
                    tempPrices[to] = prices[from] + price;
                }
            }
            swap(prices, tempPrices);
        }
        if(prices[dst] == INT_MAX){
            return -1;
        }
        return prices[dst];
    }
};
